Add CUDA support for Adagrad(fused=True)

MeetThePatel opened the pull request:
This PR adds CUDA support for Adagrad(fused=True) optimizer, along with 3 minor changes:
- Add a TensorLR variant for CPU Adagrad(fused=True).
- Fix error message in `test/test_optim.py`, where the incorrect optimizer name was being printed.
- Fix an error message in FusedSGD, where it was giving incorrect information.

Along with this PR, I have done some benchmarking for this implementation. The benchmarking script is: [gist](https://gist.github.com/MeetThePatel/b7e93f3d3b65a67a09f8b02440a8cef9). This script also has NVTX tags, so if you want to create a trace, the relevant portions will be tagged with:
- Adagrad implementation (default, foreach, fused).
- Dtype (I benchmarked across `float64`, `float32`, `float16`, and `bfloat16`).
- Tensor shape (I used tensor shapes from GPT-2 as reference).

This benchmark measures speed and memory usage.
Highlights:
- Speed: These results are not the most stable. Speedup relative to default varies wildly (sometimes being slower) based on dtype and shape, as well as from run to run. I'm not sure what's going on here.
- Memory: This fused implementation saves ~38% (peak) VRAM compared to the default and foreach implementations.

The results of this benchmark are given in the comments of the gist: [link to comment](https://gist.github.com/MeetThePatel/b7e93f3d3b65a67a09f8b02440a8cef9?permalink_comment_id=5568984#gistcomment-5568984).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

janeyx99 reviewed (commented):
Thanks for taking your time to write and benchmark fused adagrad. I have a few questions I'd like to address before reviewing the PR contents:

1. Could you move the default, foreach, fused portions of your benchmark into their own columns? Could you also clarify how speedup is calculated? A negative % speedup is not computing in my brain haha.
2. Could you also benchmark when foreach=False (this should be slowest)
3. Could you split out non related small things from this PR into another PR? Those are likely much more easily landable and it is good to have separation of distinct changes.
4. Remember to update the optim docs that fusedadagrad is supported
5. What's your plan to test correctness? I think classically we have tested that the fused CUDA + CPU versions match, and the fused + foreach=False versions match.

MeetThePatel committed:
Fix docs to show CUDA support for fused.

MeetThePatel committed:
Added fused Adagrad for CUDA

MeetThePatel commented:
> 1. Could you move the default, foreach, fused portions of your benchmark into their own columns? Could you also clarify how speedup is calculated? A negative % speedup is not computing in my brain haha.

I've updated the [gist](https://gist.github.com/MeetThePatel/b7e93f3d3b65a67a09f8b02440a8cef9) and the [comment](https://gist.github.com/MeetThePatel/b7e93f3d3b65a67a09f8b02440a8cef9?permalink_comment_id=5568984#gistcomment-5568984). I've removed the % portion, not sure how to fit that into this table format.

> 2. Could you also benchmark when foreach=False (this should be slowest)

In the new script, I've updated it to explicitly go through three configurations:
- foreach=False, fused=False
- foreach=True, fused=True
- foreach=False, fused=False

Is "foreach=False, fused=False" what you mean?

> 3. Could you split out non related small things from this PR into another PR? Those are likely much more easily landable and it is good to have separation of distinct changes.

I've split the smaller changes into the following PRs:
- https://github.com/pytorch/pytorch/pull/153074#issuecomment-2859581005
- https://github.com/pytorch/pytorch/pull/153076#issuecomment-2859590326
- https://github.com/pytorch/pytorch/pull/153078#issuecomment-2859695767

I've ripped out these commits from this branch. I'm not sure if I've done the TensorLR one correctly, as changes in this PR depend on that PR, and I'm not very adept at Git surgery.

> 4. Remember to update the optim docs that fusedadagrad is supported

Added this in [`e1076dc` (#153038)](https://github.com/pytorch/pytorch/pull/153038/commits/e1076dc98e22e4cde281560319334a60b48e7103). Are there any other locations that touch this functionality that I would need to update?

> 5. What's your plan to test correctness? I think classically we have tested that the fused CUDA + CPU versions match, and the fused + foreach=False versions match.

I had a script when I was writing the PR that I used to test correctness without launching the test suite.

I've added that to the [gist](https://gist.github.com/MeetThePatel/b7e93f3d3b65a67a09f8b02440a8cef9) as well as the [results](https://gist.github.com/MeetThePatel/b7e93f3d3b65a67a09f8b02440a8cef9?permalink_comment_id=5570048#gistcomment-5570048).

MeetThePatel committed:
Merge branch 'main' into fused-adagrad

MeetThePatel commented:
Is there anything else I should add/change to this PR?

janeyx99 reviewed (approved):
Thank you! That does make more sense!

janeyx99 commented:
Yeaaaaa you should! Thanks for seeing that; it’s to preserve FC
